### PR TITLE
LIS-5351: Rewrite all end-to-end integration tests.

### PR DESCRIPTION
**Pull request by blakesmith**
The existing test suite was either:

- Relatively mock-heavy, and not testing anything too interesting.
- Relied on locally running nsq infrastructure to complete.

Erik and I took a pass at a complete rewrite of these integration tests that test the full publisher / subscriber flows, and exercises nsqd failures and our message delivery semantics. Some other capabilities this PR enables:

1. We can simulate nsqd nodes getting disconnected from the network, and how our client responds.
2. We can now make assertions about message ordering
3. We can stand up and tear down the entire NSQ cluster between every test, for state reset
4. We can now use this to do more aggressive refactorings, including adding some round-robin publishing features we have our eye on.

We are now also requiring Java 8 and above. The client needed to support Java 6 in the past, which is now EOL.

There should be no meaningful behavior changes in the client, except for hooks that we needed to expose to make it actually testable.

Co-authored-by: Erik Helleren <erik.helleren@sproutsocial.com>

### CODE REVIEW

**Comment by stantonk** [comment]
@blakesmith @erik-helleren this is really elegant, and considering how critical nsq-j is to almost everything we do @ Sprout giving us a better test suite is 🔥.

I like the containerized approach to a meaningful end-to-end testing harness, and wrapping it in code makes it super clean and easy for others to use elsewhere. Ensuring proper state reset smooths over some of the awkwardness and poor repeatability that simply running `docker-compose up` would be prone to.

Kudos for the clever approach to testing the backoff strategy :)

Only question I had on possible other usages is how do you imagine others would use `NsqDockerCluster` in their integration tests? Some of the stuff in BaseDockerTestIT.java seems a bit tailored to your particular testing needs for this library, would they just instantiate an `NsqDockerCluster` and interact with it otherwise as if it is a "real" nsq cluster? Are there potentially helper methods or wrappers that people might want (e.g. the preservation of message ordering seems like a useful property in other folks integration tests). This is totally someday/maybe nice to have kind of stuff tho.

I'm excited to see the round-robin publishing features!

**Comment by blakesmith** [comment]
Thanks for taking a look @stantonk!

`NsqDockerCluster` was intentionally designed to be very concrete and specific to this use case. We'll probably generalize some of the ideas at some point, so that we can eventually simulate other dockerized services, and observe the outcomes in other services' test suites. Stay tuned, more to come here eventually.

As for `BaseDockerTestIT`: Yes, that will all remain specific for just the nsq-j client library tests.

File: src/main/java/com/sproutsocial/nsq/NSQMessage.java [hunk 1]
```diff
@@ -63,4 +65,16 @@ public void touch() {
         connection.touch(id);
     }
 
+    @Override
+    public void forceFlush() {
```
Comment (ravisanwal) [inline src/main/java/com/sproutsocial/nsq/NSQMessage.java:69]: What are the semantics of this method.
It looks like it is doing an operation on connection and is not scoped to the current message.
Is connection/subscriber a better place for this?

File: src/test/java/com/sproutsocial/nsq/BackoffDockerTestIT.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
+package com.sproutsocial.nsq;
+
+import org.junit.Test;
```
Comment (ravisanwal) [inline src/test/java/com/sproutsocial/nsq/BackoffDockerTestIT.java:3]: This is a NIT: may be we can upgrade to junit5. Can be done in a later change as well though.

File: src/test/java/com/sproutsocial/nsq/BaseDockerTestIT.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        for (NSQMessage e : receivedMessages) {
+            Assert.assertEquals(cluster.getNsqdNodes().get(nsqHostIndex).getTcpHostAndPort(), e.getConnection().getHost());
+        }
+    }
+
+    protected Publisher primaryOnlyPublisher() {
+        return new Publisher(client, cluster.getNsqdNodes().get(0).getTcpHostAndPort().toString(), null);
+    }
+
+    protected Publisher backupPublisher() {
```
Comment (ravisanwal) [inline src/test/java/com/sproutsocial/nsq/BaseDockerTestIT.java:110]: NIT: This is not a backup publisher but a publisher with backup NSQD, so may be `publisherWithBackup` as a name?

**Review by ravisanwal — COMMENTED** [review]
This is great.
I really like the code-contained style of integration tests.
I have some minor comments and a clarification question.

Other that those, this looks great.

File: src/main/java/com/sproutsocial/nsq/NSQMessage.java (hunk 1 above)
Comment (erik-helleren) [reply src/main/java/com/sproutsocial/nsq/NSQMessage.java:69]: We don't have direct access to the connection from the consumer side.  

This is being added to support strongly consistent finishes.  This speeds up tests and makes them deterministic.

**Review by erik-helleren — APPROVED** [review]

**Pull request merged**